### PR TITLE
Refactor - Sanitization of memory accesses in JIT

### PR DESCRIPTION
**Pull request by Lichtso**

### CODE REVIEW

File: src/jit.rs [hunk 1]
```diff
@@ -200,7 +200,7 @@ const ANCHOR_INTERNAL_FUNCTION_CALL_PROLOGUE: usize = 12;
 const ANCHOR_INTERNAL_FUNCTION_CALL_REG: usize = 13;
 const ANCHOR_CALL_REG_UNSUPPORTED_INSTRUCTION: usize = 14;
 const ANCHOR_TRANSLATE_MEMORY_ADDRESS: usize = 21;
-const ANCHOR_COUNT: usize = 30; // Update me when adding or removing anchors
+const ANCHOR_COUNT: usize = 34; // Update me when adding or removing anchors
```
Comment (LucasSte) [inline src/jit.rs:203]: You haven't added or removed anchors, have you?

File: src/jit.rs [hunk 2]
```diff
@@ -873,29 +876,24 @@ impl<'a, C: ContextObject> JitCompiler<'a, C> {
...
-            let key = self.diversification_rng.gen::<i32>() as i64;
-            self.emit_ins(X86Instruction::load_immediate(destination, value.rotate_right(32).wrapping_sub(key)));
-            self.emit_ins(X86Instruction::alu(OperandSize::S64, 0x81, 0, destination, key, None)); // wrapping_add(key)
+            self.emit_ins(X86Instruction::load_immediate(destination, value.rotate_right(32).wrapping_sub(lower_key)));
+            self.emit_ins(X86Instruction::alu(OperandSize::S64, 0x81, 0, destination, lower_key, None)); // wrapping_add(lower_key)
             self.emit_ins(X86Instruction::alu(OperandSize::S64, 0xc1, 4, destination, 32, None)); // shift_left(32)
+        } else if destination != REGISTER_SCRATCH {
+            self.emit_ins(X86Instruction::load_immediate(destination, value.wrapping_sub(self.immediate_value_key)));
+            self.emit_ins(X86Instruction::load_immediate(REGISTER_SCRATCH, self.immediate_value_key));
+            self.emit_ins(X86Instruction::alu(OperandSize::S64, 0x01, REGISTER_SCRATCH, destination, 0, None));
```
Comment (LucasSte) [inline src/jit.rs]: I like the comments you add for clarity next to instructions. One could be useful here for signaling an addition.

File: src/jit.rs [hunk 3]
```diff
@@ -1190,8 +1184,12 @@ impl<'a, C: ContextObject> JitCompiler<'a, C> {
         }
 
         if self.config.enable_address_translation {
-            let access_type = if value.is_none() { AccessType::Load } else { AccessType::Store };
-            let anchor = ANCHOR_TRANSLATE_MEMORY_ADDRESS + len.trailing_zeros() as usize + 4 * (access_type as usize);
+            let access_type = match value {
```
Comment (LucasSte) [inline src/jit.rs]: Calling this `access_type` may be confusing with `AccessType`. What about altering the enum to represent load, store reg and store imm?

File: src/jit.rs [hunk 4]
```diff
@@ -1627,7 +1623,11 @@ impl<'a, C: ContextObject> JitCompiler<'a, C> {
                     Argument { index: 1, value: Value::RegisterPlusConstant32(REGISTER_PTR_TO_VM, self.slot_in_vm(RuntimeEnvironmentSlot::MemoryMapping), false) },
                     Argument { index: 0, value: Value::RegisterPlusConstant32(REGISTER_PTR_TO_VM, self.slot_in_vm(RuntimeEnvironmentSlot::ProgramResult), false) },
                 ], None);
-            } else {
+            } else { // AccessType::Store
+                if *access_type == 8 {
+                    // Second half of emit_sanitized_load_immediate()
+                    self.emit_ins(X86Instruction::alu(OperandSize::S64, 0x81, 0, RSP, lower_key, Some(X86IndirectAccess::OffsetIndexShift(-96, RSP, 0))));
```
Comment (LucasSte) [inline src/jit.rs:1629]: How about adding a comment about the -96 offset? Does it come from `stack_slot_of_value_to_store`?

File: src/jit.rs [hunk 5]
```diff
@@ -1650,8 +1650,10 @@ impl<'a, C: ContextObject> JitCompiler<'a, C> {
             self.emit_ins(X86Instruction::xchg(OperandSize::S64, REGISTER_SCRATCH, RSP, Some(X86IndirectAccess::OffsetIndexShift(0, RSP, 0)))); // Swap return address and self.pc
             self.emit_ins(X86Instruction::conditional_jump_immediate(0x85, self.relative_to_anchor(ANCHOR_THROW_EXCEPTION, 6)));
 
-            // unwrap() the result into REGISTER_SCRATCH
-            self.emit_ins(X86Instruction::load(OperandSize::S64, REGISTER_PTR_TO_VM, REGISTER_SCRATCH, X86IndirectAccess::Offset(self.slot_in_vm(RuntimeEnvironmentSlot::ProgramResult) + std::mem::size_of::<u64>() as i32)));
+            if *access_type == 0 { // AccessType::Load
```
Comment (LucasSte) [inline src/jit.rs]: Why is this restricted to load?

Comment (Lichtso) [reply src/jit.rs]: Only load instructions do return a value, store instructions don't.

File: src/jit.rs (hunk 1 above)
Comment (Lichtso) [reply src/jit.rs:203]: The store immediate ones have their own ANCHOR_TRANSLATE_MEMORY_ADDRESS now.

File: src/jit.rs (hunk 3 above)
Comment (Lichtso) [reply src/jit.rs]: `AccessType` comes from the memory mapping and that does not know about store immediate.
So I renamed `access_type` to `anchor_base`.

**Review by LucasSte — APPROVED** [review]

**Pull request merged**